Python: Add Streamlit models

sylwia-budzynska opened the pull request:
👋 This PR adds a few sources and a sink for the [Streamlit](https://streamlit.io/) framework. Streamlit is a popular Python library for creating web apps for machine learning and data science. 🚀 

I did a bit of tweaking with tests, let me know if these could be improved.

Side note: Streamlit creates SQL connection to databases using SQL Alchemy Engine. There are a few potential SQLExecution sinks that I haven't included, because to execute the SQL query they require sqlalchemy's `TextClause` object, which is already modeled, so my assumption here is that we don't need to model those.

For example `st.connection().session.execute()` would be a potential sink, but it requires a `TextClause` object, which is already modeled as a SQL injection sink. 

Here is an example of how it could be used:
```python
import streamlit as st
from sqlalchemy import text

# Initialize connection.
conn = st.connection("postgresql", type="sql")

inp = st.text_input("Query the database")

#BAD
if st.button("Add the number!"):
    with conn.session as session:
        query = text(f'SELECT {inp} FROM mytable;')
        for value in session.execute(query):
            st.write(value)
        session.commit()
```
Note that `text()` is not used in Streamlit's documention, but if we run the snippet from the docs (https://docs.streamlit.io/develop/api-reference/connections/st.connections.sqlconnection#:~:text=SQLConnection.engine-,SQLConnection.session,-Streamlit%20Version), it returns an error `ArgumentError: Textual SQL expression 'INSERT INTO mytable (val)...’ should be explicitly declared as text(‘INSERT INTO mytable (val)...’)` - it looks like Streamlit docs haven't been updated.

If you think we should model this, please let me know.

yoff reviewed (changes requested):
Looks generally good. I think we distinguish between sql _executions_, which is what the streamlit `query` function represents, and then sql _constructions_, which is what `TextClause` represents. It might be good to add the connection to SqlAlchemy; I have made a suggestion to that effect.

The file `rfs-test.ql` looks like it could be generally useful. We might want to put it somewhere central and simply import it, as we do for `ConceptTest`. It does not have to be in this PR, though.

sylwia-budzynska commented on python/ql/lib/semmle/python/frameworks/Streamlit.qll:
```diff
@@ -0,0 +1,47 @@
+
+    override DataFlow::Node getSql() { result in [this.getArg(0), this.getArgByName("sql")] }
+  }
+}
```
This is a very similar case to  the SQL Alchemy session `streamlit.connection().session.execute()` that I mentioned in the description.  In this case, should we also model `streamlit.connection().session.execute()` as `SqlAlchemy::Session::InstanceSource` in addition to the Connection you've modeled here? 

(Side note that the methods for `streamlit.connection().connect()` (for example `execute()`) also require a `TextClause` object to be executed. The Streamlit docs are wrong here too.)

```python
import streamlit as st
from sqlalchemy import text

conn = st.connection("postgresql", type="sql").connect()

inp = st.text_input("Query the database")

#BAD
with conn:
    for value in conn.execute(text(f'SELECT {inp} FROM mytable;')):
        st.write(value)
```

yoff replied:
You are right, sessions are not covered by PEP 249, so we should extend `SqlAlchemy::Session::InstanceSource` in a similar way. You could do the same with `SqlAlchemy::Engine::InstanceSource` and `streamlit.connection().engine`, actually. Then you will have covered a wild range of uses.. :-)

yoff replied:
I think we will have to leave it to the queries to distinguish between executinos and constructions, in the library models, we can mark what is what. If the code can run, and it follows the pattern we are looking for, alerting should be fine.

yoff commented:
> > The file rfs-test.ql looks like it could be generally useful. We might want to put it somewhere central and simply import it, as we do for ConceptTest. It does not have to be in this PR, though.
> 
> This is actually copied from what you [wrote](https://github.com/github/codeql/pull/16135#discussion_r1594731326) in my last PR, so all credit goes to you 😄

Ah, right. I guess it is the second time you are testing remote sources for a library, then. We should just make it a one-line import.

sylwia-budzynska commented:
>

> > > The file rfs-test.ql looks like it could be generally useful. We might want to put it somewhere central and simply import it, as we do for ConceptTest. It does not have to be in this PR, though.
> > 
> > 
> > This is actually copied from what you [wrote](https://github.com/github/codeql/pull/16135#discussion_r1594731326) in my last PR, so all credit goes to you 😄
> 
> Ah, right. I guess it is the second time you are testing remote sources for a library, then. We should just make it a one-line import.

AFAIK the test you've helped me write for Gradio remote flow sources has never been centralized, and so it sits in the [Gradio framework tests](https://github.com/sylwia-budzynska/codeql/blob/main/python/ql/test/library-tests/frameworks/gradio/source_test.ql). As far as I can see it's not in [`ConceptsTest`](https://github.com/github/codeql/blob/main/python/ql/test/experimental/meta/ConceptsTest.qll). Or do you mean a different import?

yoff commented:
> AFAIK the test you've helped me write for Gradio remote flow sources has never been centralized, and so it sits in the [Gradio framework tests](https://github.com/sylwia-budzynska/codeql/blob/main/python/ql/test/library-tests/frameworks/gradio/source_test.ql). As far as I can see there are no remote flow source tests in [`ConceptsTest`](https://github.com/github/codeql/blob/main/python/ql/test/experimental/meta/ConceptsTest.qll). Or do you mean a different import?

I meant that we should write a new `.qll`-file, similar to `ConceptTest.qll`, containing the remote flow source test. Since this is something you frequently want to test, we should make it easy. We might also be cheeky and simply add it to `ConceptTest.qll` since it is similar to a concept...

sylwia-budzynska commented:
@yoff I moved remoteflowsource test and added session, connection and engine models.

I wrote a new class, [StreamlitSqlConnection](https://github.com/github/codeql/pull/17080/files#diff-2ef290fa6cf6b8f242a8a57d839b49670f5d3ee6134d38fbd2142034f30cb194R39), to differentiate between connections to SQL databases, and Snowflake and Snowpark databases. The last two are based on the snowflake-connector-python library, which is something we might want to model at a later date.

We can connect to SQL databases for example with `import streamlit as st; conn = st.connection(‘pets_db’, type=‘sql’)`, where “sql” can be replaced with “snowflake” or “snowpark” for these databases. The StreamlitSqlConnection filters for only the cases with “sql”.

yoff commented on python/ql/lib/semmle/python/frameworks/Streamlit.qll:
```diff
@@ -8,6 +8,7 @@ import semmle.python.dataflow.new.RemoteFlowSources
 import semmle.python.dataflow.new.TaintTracking
 import semmle.python.ApiGraphs
 import semmle.python.Concepts
+import semmle.python.frameworks.SqlAlchemy
```
```suggestion
private import semmle.python.frameworks.SqlAlchemy
```
make this private to remove the warning (we are not interested in re-exporting it anyway).

yoff reviewed (approved):
Very nice changes! We will want to make the Gradio tests also use the new shared module, but that does not have to be in this PR.

sylwia-budzynska commented:
I updated the Gradio test that should use the module! Should be good to go now 🚀

yoff reviewed (approved):
Excellent :-)

sylwia-budzynska commented:
I've never done a DCA run before, but it looks fairly straightforward, so I'll try. I'll ask for assistance if anything pops up

yoff commented:
DCA run looks fine :-)
